### PR TITLE
Calls to torch/daml from MLBalance

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
It is still a plumbing PR, absolutely zero science yet. The results from the 3dvar test are pretty bad, but I'm blaming this on:
- The nn being trained on a 1/4 deg ocean/ice background and for a different season.
- I invented values for the missing variables
- the diag of B is 1 for all variables ... 

**Work done:** 
- MLBalance is only built and tested if `daml` and `torch` are found
- Use of the gradient of the ice emulator from `daml` to initialize the Jacobian in `MLBalance`.
- git-lfs storage of the weights as binary files for the 2 poles. They came from training the nn with a 1/4 deg background. The gradient of the emulator looks reasonable when the inference is done at the same resolution and with a background close enough to the training date ... probably clear as mud. 
- I invented the background for the missing variables (Tair, Tsfc and sice) and updated the `cice` restart/background file 

### Issue(s) addressed
- fixes #1015 

## Testing

How were these changes tested?  **a dirac and a 3dvar test were added**
What compilers / HPCs was it tested with?  **gnu only**
Are the changes covered by ctests? **Yes and no, mlbalance and it associated tests are only tested if daml/torch are found**

## Dependencies
`torch` & `daml`

### CODE REVIEW

File: src/soca/MLBalance/MLBalance.cc [hunk 1]
```diff
@@ -37,6 +38,9 @@ MLBalance::MLBalance(
     innerGeometryData_(outerGeometryData),
     jac_(nullptr)
 {
+  // setup geometry
+  geom_.reset(new Geometry(params_.geometry.value(), outerGeometryData.comm()));
```
Comment (travissluka) [inline src/soca/MLBalance/MLBalance.cc]: Don't create a new geometry.
I think you can pull the data you need from `outerGeometryData.functionSpace()` which has the lat/lon information.

File: src/soca/MLBalance/MLBalance.h [hunk 2]
```diff
@@ -48,11 +48,16 @@ class MLBalance : public saber::SaberOuterBlockBase {
...
+  void setupJac(const oops::FieldSet3D &,
+                const eckit::mpi::Comm &,
+                const eckit::Configuration &);
+
  private:
   void print(std::ostream &) const override;
   oops::Variables activeVars_;
   oops::Variables innerVars_;
   const oops::GeometryData & innerGeometryData_;
+  std::shared_ptr<Geometry> geom_;
```
Comment (travissluka) [inline src/soca/MLBalance/MLBalance.h]:
```suggestion
```

see above comment

File: src/soca/MLBalance/MLBalanceParameters.h [hunk 3]
```diff
@@ -20,7 +20,7 @@ class MLBalanceParameters : public saber::SaberBlockParametersBase {
   OOPS_CONCRETE_PARAMETERS(MLBalanceParameters, saber::SaberBlockParametersBase)
  public:
   oops::RequiredParameter<eckit::LocalConfiguration> mlbalances{"ML Balances", this};
-
+  oops::RequiredParameter<eckit::LocalConfiguration> geometry{"geometry", this};
```
Comment (travissluka) [inline src/soca/MLBalance/MLBalanceParameters.h]:
```suggestion
```

File: src/soca/MLBalance/MLBalance.cc [hunk 4]
```diff
@@ -52,12 +56,10 @@ MLBalance::MLBalance(
   }
 
   // Initialize the Jacobian
-  // TODO(G): Currently set to 1.0 for testing, reset to 0 before implementing a
-  //           realistic jacobian
-  jac_ = util::createFieldSet(xb["tocn"].functionspace(), jacVars, 1.0);
+  jac_ = util::createFieldSet(xb["tocn"].functionspace(), jacVars, 0.0);
```
Comment (travissluka) [inline src/soca/MLBalance/MLBalance.cc]:
```suggestion
  jac_ = util::createFieldSet(outerGeometryData.functionspace(), jacVars, 0.0);
```

File: src/soca/MLBalance/MLJac.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        dcdhi(jnode, 0) = dcdx[5].item<float>();
+      }
+      oops::Log::trace() << "In MLJack" << std::endl;
+    }
+
+    const eckit::LocalConfiguration getConf(const eckit::Configuration & conf,
+                                            std::string str) {
+      const eckit::LocalConfiguration localConf(conf, str);
+      return localConf;
+    }
```
Comment (travissluka) [inline src/soca/MLBalance/MLJac.h:87]: why is this needed?

**Review by travissluka — CHANGES_REQUESTED** [review]

File: src/soca/MLBalance/MLJac.h (hunk 5 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/MLJac.h:87]: ~left over from trial and error probably ... I'll remove.~
It's used in the nn initializer, I couldn't find a function that would return a "sub-configuration" from a config. Is there a better way to do this?

File: src/soca/MLBalance/MLBalance.cc (hunk 1 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/MLBalance.cc]: removed and using `functionsSpace.lonlat()` instead.

**Comment by travissluka** [comment]
oops, I added conflicts!

**Comment by shlyaeva** [comment]
I've been "ignoring" this because I saw there were some comments from Travis and some conflicts; let me know when ready for review.

**Comment by guillaumevernieres** [comment]
> I've been "ignoring" this because I saw there were some comments from Travis and some conflicts; let me know when ready for review.

It's ready to review @shlyaeva . I don't think the ci failures are related to this work.

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -650,7 +666,26 @@ soca_add_test( NAME 3dvar_nicas
...
-## TODO(travis) fix the lowres grid dx,dy so that I can init the 
+if( ${daml_FOUND} )
+  soca_add_test( NAME 3dvar_diffmlb
+                 EXE  soca_var.x
+                 TEST_DEPENDS test_soca_gridgen
+                              test_soca_parameters_diffusion_hz
+                              test_soca_parameters_diffusion_vt)
+endif()
+
+soca_add_test( NAME 3dvar_godas
```
Comment (travissluka) [inline test/CMakeLists.txt]: @guillaumevernieres bad merge of develop. check this file again and put some things back the way they were

Comment (guillaumevernieres) [reply test/CMakeLists.txt]: ~Just fixed yesterday's sloppy merge.~

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -17,6 +17,9 @@ set( soca_test_input
   testinput/3dhyb_diffusion.yml
   testinput/3dhyb.yml
   testinput/3dhybfgat.yml
+  testinput/3dvar_diffusion.yml
+  testinput/3dvar_diffmlb.yml
+  testinput/3dvar_godas.yml
```
Comment (travissluka) [inline test/CMakeLists.txt]:
```suggestion
```

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -650,7 +665,15 @@ soca_add_test( NAME 3dvar_nicas
                TEST_DEPENDS test_soca_parameters_bump_cor_nicas )
 
 
-## TODO(travis) fix the lowres grid dx,dy so that I can init the 
+if( ${daml_FOUND} )
+  soca_add_test( NAME 3dvar_diffmlb
+                 EXE  soca_var.x
+                 TEST_DEPENDS test_soca_gridgen
+                              test_soca_parameters_diffusion_hz
+                              test_soca_parameters_diffusion_vt)
```
Comment (travissluka) [inline test/CMakeLists.txt]:
```suggestion
                              test_soca_parameters_diffusion )
```

File: test/testinput/3dvar_diffmlb.yml [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+      write increment: true
+      increment:
+        state component:
+          datadir: data_generated/3dvar_diffmlb
+          date: *bkg_date
+          exp: 3dvar_diffmlb.iter1
+          type: incr
+
+output:
+  datadir: data_generated/3dvar_diffmlb
```
Comment (travissluka) [inline test/testinput/3dvar_diffmlb.yml]:
```suggestion
  datadir: data_output/
```

File: test/testinput/3dvar_diffmlb.yml [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  exp: 3dvar_diffmlb
+  type: an
+
+final:
+  diagnostics:
+    departures: oman
+  increment:
+    output:
+      state component:
+        datadir: data_generated/3dvar_diffmlb
```
Comment (travissluka) [inline test/testinput/3dvar_diffmlb.yml]:
```suggestion
        datadir: data_output/
```

File: test/testinput/3dvar_diffmlb.yml [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+  observations:
+    observers:
+
+    - obs space:
+        name: SeaIceFraction
+        obsdataout:
+          engine:
+            type: H5File
+            obsfile: data_generated/3dvar_diffmlb/icec.3dvar_diffmlb.nc
```
Comment (travissluka) [inline test/testinput/3dvar_diffmlb.yml]:
```suggestion
            obsfile: data_output/icec.3dvar_diffmlb.nc
```

File: test/CMakeLists.txt (hunk 7 above)
Comment (travissluka) [reply test/CMakeLists.txt]: still a sloppy merge!

**Review by travissluka — CHANGES_REQUESTED** [review]
minor change request.
For the record though, I don't like the dependency on `daml`. I would like that eliminated once we have torch part of spack stack and are able to exercise this stuff in our CI

**Comment by guillaumevernieres** [comment]
> minor change request. For the record though, I don't like the dependency on `daml`. I would like that eliminated once we have torch part of spack stack and are able to exercise this stuff in our CI

Will do. I don't think I have a choice anyway. I'm facing circular dependencies that I don't know how to resolve, other than by migrating the code inside soca.

**Review by travissluka — APPROVED** [review]

File: src/soca/MLBalance/MLBalance.cc [hunk 12]
```diff
@@ -66,6 +65,16 @@ MLBalance::~MLBalance() {}
 
 // --------------------------------------------------------------------------------------
 
+void MLBalance::setupJac(const oops::FieldSet3D & xb,
+                         const eckit::mpi::Comm & comm,
+                         const eckit::Configuration & config) {
```
Comment (shlyaeva) [inline src/soca/MLBalance/MLBalance.cc:70]: Nitpick: looks like this argument is not used and can be removed.

File: src/soca/MLBalance/MLJac.h (hunk 5 above)
Comment (shlyaeva) [reply src/soca/MLBalance/MLJac.h:87]: `conf.getSubConfiguration(str)`?

**Review by shlyaeva — APPROVED** [review]
Neat!

**Pull request merged**